### PR TITLE
Rs 8392 (merging by variable)

### PR DESCRIPTION
**Pull request by JustinCCYap**

### CODE REVIEW

File: tests/testthat/test-mergingandstackingutilities.R [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    categorical.var <- structure(c(1, 2, 1),
+                                 labels = structure(1:2, .Names = c("A", "B")))
+    expect_equal(variableType(variable = categorical.var), "Categorical")
+    dates <- structure(c(18802, 18803, 18804), class = "Date")
+    expect_equal(variableType(variable = dates), "Date")
+    date.times <- structure(c(1624538040, 1624589160, 1624733100),
+                            class = c("POSIXct", "POSIXt"), tzone = "UTC")
+    expect_equal(variableType(variable = date.times), "Date/Time")
+    duration <- structure(10:12, class = "difftime", units = "mins")
+    expect_equal(variableType(variable = duration), "Duration")
```
Comment (oliverbock) [inline tests/testthat/test-mergingandstackingutilities.R:13]: You should probably change these to user the new constants.  Also I think I saw them in the main file.

File: tests/testthat/test-mergingandstackingutilities.R [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+test_that("splitByComma", {
+    expect_equal(splitByComma(input.text = "Q1,Q2"), c("Q1", "Q2"))
+    expect_equal(splitByComma(input.text = ",Q1,Q2, ,,Q3,"), c("Q1", "Q2", "Q3"))
+    expect_equal(splitByComma(input.text = ",Q1(,2, 3,,), ,, Q2(3,4),,Q3,",
+                              ignore.commas.in.parentheses = TRUE),
+                 c("Q1(,2, 3,,)", "Q2(3,4)", "Q3"))
+})
+
+test_that("isIntegerValued", {
+    expect_true(isIntegerValued(x = c(-1, 0, 1, 2, 3, NA, Inf)))
```
Comment (oliverbock) [inline tests/testthat/test-mergingandstackingutilities.R:39]: Infinity is an integer?  That cuts things down!  We might yet isolate its real value :-)

File: tests/testthat/test-mergedatasetsbyvariable.R [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+#             with "Low self-monitor" cases removed.
+# cola17.sav: Q1_* to Q4_* included, with only "High self-monitor" cases kept.
+# cola18.sav: Q1_* to Q4_* included, with IDs in Attr1 (renamed as Attr1_dup)
+#             modified to include duplicate values.
+
+test_that("Example used for widget test in flipFormat", {
+    expect_error(merge.data.set.by.var.output <- MergeDataSetsByVariable(data.set.names = c(findInstDirFile("cola15.sav"),
+                                                                                            findInstDirFile("cola16.sav")),
+                                                                         id.variables = c("Attr1","PartyID"),
+                                                                         variables.to.include.or.omit = list("Q1_A_c", "Q4_B")), NA)
```
Comment (oliverbock) [inline tests/testthat/test-mergedatasetsbyvariable.R:20]: Surely there is a better way to indicate that a function should return no error, than by a call to something called "expect error"!  Perhaps `expect_silent()`?  Perhaps write a `function expect_no_error(f) { expect_error(f, NA) }`  (A naive reading of an `expect_error(..., NA)` is that an error is expected, but its exact text/type is unspecified.)

**Review by oliverbock — APPROVED** [review]
Now I admit that this was shorter, but it was much easier to read.  Thanks!

Also thanks for the helpful comments on tricky bits of code.

File: R/mergedatasetsbyvariable.R [hunk 4]
```diff
@@ -0,0 +1,601 @@
+#' @title Merge Data Sets by Variable
+#' @description Merges multiple data sets by case where the data sets contain
+#'  similar variables but different cases, e.g., data sets from different time
+#'  periods.
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: Fixed in 5ab565c

File: R/mergedatasetsbyvariable.R [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+#' @param include.or.omit.variables A character vector where each element
+#'  corresponds to an input data set, and indicates whether variables from the
+#'  input data set are to be specified in the merged data set by specifying
+#'  the variables to include ("Only include manually specified variables") or the
+#'  variables to omit ("Include all variables except those manually omitted").
+#' @param variables.to.include.or.omit A list of character vectors corresponding
+#'  to each data set. Each character vector contains comma-separated names of
+#'  variables to include or omit (depending on the option for the data set in
+#'  \code{include.or.omit.variables}). Ranges of variables can be specified by
+#'  separating variable names by '-'.
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: This reflects the UI: there is at least one textbox for each data set, and each text box contains comma-separated names. I'll improve the wording of this.

File: R/mergedatasetsbyvariable.R [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+    #                              data sets. Each element contains the names of
+    #                              variables from a data set that were omitted
+    #                              from the merged data set.
+    # merged.id.variable.name: A character scalar of the name of the ID
+    #                          variable in the merged data set. It is NULL if
+    #                          there is no ID variable.
+    # source.data.set.indices: An integer vector corresponding to the variables
+    #                          in the merged data set. Each element contains
+    #                          the index of the input data set from which the
+    #                          variable originated.
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: The first data set index is used. I have updated the documentation to mention this in 2fd7df8

File: R/mergedatasetsbyvariable.R [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+    #                     variables from the input data sets.
+    # merged.data.set.var.names: A character vector containing the names of the
+    #                            variables in the merged data set. Has the
+    #                            attributes input.variable.names.list,
+    #                            omitted.variable.names, merged.id.variable.name,
+    #                            source.data.set.indices.
+    # input.variable.names.list: A list of character vectors corresponding to
+    #                            the input data sets. Each element contains the
+    #                            names of the variables from an input data set
+    #                            that will be present in the merged data set.
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: OK, name changed in 2fd7df8

File: R/mergedatasetsbyvariable.R [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+    matched.cases.matrix <- matchCases(input.data.sets.metadata, id.variables,
+                                       data.sets,
+                                       only.keep.cases.matched.to.all.data.sets)
+    merged.data.set.var.names <- mergedDataSetVariableNames(input.data.sets.metadata,
+                                                            include.or.omit.variables,
+                                                            variables.to.include.or.omit,
+                                                            matched.cases.matrix)
+    merged.data.set <- mergedDataSetByVariable(data.sets, matched.cases.matrix,
+                                               merged.data.set.var.names,
+                                               input.data.sets.metadata)
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: OK, name changed to doMergeByVariable in 2fd7df8

File: R/mergedatasetsbyvariable.R [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+    result$source.data.set.indices <- attr(merged.data.set.var.names,
+                                           "source.data.set.indices")
+    result$omitted.variable.names.list <- attr(merged.data.set.var.names,
+                                               "omitted.variable.names.list")
+    result$merged.id.variable.name <- attr(merged.data.set.var.names,
+                                           "merged.id.variable.name")
+    result$id.variable.names <- attr(matched.cases.matrix, "id.variable.names")
+    result$example.id.values <- exampleIDValues(result$id.variable.names,
+                                                data.sets)
+    result$is.saved.to.cloud <- IsDisplayrCloudDriveAvailable()
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: Yep, I'm now setting is.saved.to.cloud first then passing it to writeDataSet()

File: R/mergedatasetsbyvariable.R [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+#'  be converted to.
+#' @return A vector representing the converted ID variable.
+#' @noRd
+convertIDVariableType <- function(ids, id.variable.type,
+                                  merged.id.variable.type)
+{
+    if (merged.id.variable.type == DATE.TIME.VARIABLE.TYPE &&
+        id.variable.type == DATE.VARIABLE.TYPE)
+        AsDateTime(as.character(ids))
+    else # merged.id.variable.type == TEXT.VARIABLE.TYPE
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: No it is correct but perhaps not obvious. I have refactored it slightly and improved the comment.

File: R/mergedatasetsbyvariable.R [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+        if (id.variable.type == CATEGORICAL.VARIABLE.TYPE)
+        {
+            converted.ids <- rep(NA_character_, length(ids))
+            val.attrs <- attr(ids, "labels")
+            val.labels <- names(val.attrs)
+            for (i in seq_along(val.attrs))
+                converted.ids[ids == val.attrs[i]] <- val.labels[i]
+            converted.ids
+        }
+        else
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R:339]: Numeric IDs are allowed as long as all ID variables are numeric. Hopefully this should be clearer with my changes.

File: R/mergedatasetsbyvariable.R [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+            if (all(!(non.missing.ids.list[[i]] %in% unlist(non.missing.ids.list[-i]))))
+                warning("The IDs in data set ", i, " from variable '",
+                        id.var.names[i],
+                        "' are not present in the other data sets. ",
+                        "Ensure that the ID variable names have been correctly specified.")
+        }
+    }
+
+    # Merge ID values
+    unique.ids <- removeNA(unique(unlist(non.missing.ids.list)))
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/mergedatasetsbyvariable.R]: Yes, I have removed it in 2fd7df8

File: tests/testthat/test-mergingandstackingutilities.R (hunk 1 above)
Comment (JustinCCYap) [reply tests/testthat/test-mergingandstackingutilities.R]: Done in 2fd7df8

File: tests/testthat/test-mergingandstackingutilities.R (hunk 2 above)
Comment (JustinCCYap) [reply tests/testthat/test-mergingandstackingutilities.R]: Oops, I have changed the function to not recognise Inf as integer in 2fd7df8

File: tests/testthat/test-mergedatasetsbyvariable.R (hunk 3 above)
Comment (JustinCCYap) [reply tests/testthat/test-mergedatasetsbyvariable.R]: Created function expectNoError. The documentation for expect_error says that when NA is passed in, no error is expected.

**Pull request merged**